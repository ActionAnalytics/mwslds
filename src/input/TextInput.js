import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { startCase } from 'lodash'

import '../style'

const propTypes = {
  name: PropTypes.string,
  title: PropTypes.string,
  value: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  prefix: PropTypes.string,
  disabled: PropTypes.bool,
  readOnly: PropTypes.bool,
  children: PropTypes.element,
  width: PropTypes.string,
  validationError: PropTypes.string,
  placeholder: PropTypes.string,
}

const defaultProps = {
  value: undefined,
  name: '',
  title: null,
  prefix: null,
  disabled: null,
  readOnly: null,
  children: null,
  width: null,
  validationError: null,
  placeholder: undefined,
}

function onChangeWrapper(wrapped) {
  return evt => wrapped(evt.target.value)
}

function TextInput(props) {
  const {
    name,
    title,
    value,
    onChange,
    prefix,
    disabled,
    readOnly,
    width,
    validationError,
    placeholder,
  } = props

  const inputId = `${prefix || ''}${name}`
  const labelClasses = classnames('control-label', validationError && 'has-error')

  return (
    <label className={labelClasses} htmlFor={inputId} style={{ width }}>
      {title || startCase(name)}
      <input
        id={inputId}
        className="form-control"
        value={value}
        placeholder={placeholder}
        onChange={onChangeWrapper(onChange)}
        type="text"
        disabled={disabled}
        readOnly={readOnly}
      />
      {props.children}
    </label>
  )
}

TextInput.propTypes = propTypes
TextInput.defaultProps = defaultProps

export default TextInput
