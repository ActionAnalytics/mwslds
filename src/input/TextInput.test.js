import React from 'react'
import { configure, shallow, mount } from 'enzyme'
import { createSerializer } from 'enzyme-to-json'
import Adapter from 'enzyme-adapter-react-16'

import TextInput from './TextInput'

configure({ adapter: new Adapter() })
expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }))

describe('TextInput', () => {
  it('renders', () => {
    const text = shallow(<TextInput name="texty" onChange={() => {}} />)
    expect(text).toMatchSnapshot()
  })

  it('has the correct title', () => {
    const text = shallow(<TextInput name="texty" title="not texty" onChange={() => {}} />)
    expect(text).toMatchSnapshot()
  })

  it('has the correct width', () => {
    const wide = shallow(<TextInput name="texty" width="100%" onChange={() => {}} />)
    expect(wide).toMatchSnapshot()

    const lesswide = shallow(<TextInput name="texty" width="50%" onChange={() => {}} />)
    expect(lesswide).toMatchSnapshot()
  })

  it('displays the correct value', () => {
    const hello = mount(<TextInput name="texty" value="hello" onChange={() => {}} />)
    expect(hello).toMatchSnapshot()

    const empty = mount(<TextInput name="texty" value="" onChange={() => {}} />)
    expect(empty.find('input').props().value).toBe('')
  })

  it('returns value for change events', () => {
    const text = shallow(<TextInput name="texty" value="whatever" onChange={val => val} />)
    const input = text.find('input')
    expect(input.props().onChange({ target: { value: 'hi' } })).toBe('hi')
    expect(input.props().onChange({ target: { value: null } })).toBe(null)
  })

  it('is disabled properly', () => {
    const disabled = shallow(<TextInput name="texty" disabled onChange={() => {}} />)
    expect(disabled.find('input').props().disabled).toBe(true)
  })

  it('is readOnly properly', () => {
    const readOnly = shallow(<TextInput name="texty" readOnly onChange={() => {}} />)
    expect(readOnly.find('input').props().readOnly).toBe(true)
  })

  it('shows placeholder properly', () => {
    const nice = mount(<TextInput name="texty" placeholder="nice" onChange={() => {}} />)
    expect(nice).toMatchSnapshot()
    // checks like this are redundant but useful if the snapshot gets messed up
    expect(nice.find('input').props().placeholder).toBe('nice')
  })

  it('shows errors properly', () => {
    const error = mount(<TextInput name="texty" validationError="whatever" onChange={() => {}} />)
    expect(error).toMatchSnapshot()
    expect(error.find('.has-error').length).toBe(1)
  })
})
