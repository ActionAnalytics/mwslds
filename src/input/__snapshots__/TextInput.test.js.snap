// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput displays the correct value 1`] = `
<label
  className="control-label"
  htmlFor="texty"
  style={
    Object {
      "width": null,
    }
  }
>
  Texty
  <input
    className="form-control"
    disabled={null}
    id="texty"
    onChange={[Function]}
    readOnly={null}
    type="text"
    value="hello"
  />
</label>
`;

exports[`TextInput has the correct title 1`] = `
<label
  className="control-label"
  htmlFor="texty"
  style={
    Object {
      "width": null,
    }
  }
>
  not texty
  <input
    className="form-control"
    disabled={null}
    id="texty"
    onChange={[Function]}
    readOnly={null}
    type="text"
  />
</label>
`;

exports[`TextInput has the correct width 1`] = `
<label
  className="control-label"
  htmlFor="texty"
  style={
    Object {
      "width": "100%",
    }
  }
>
  Texty
  <input
    className="form-control"
    disabled={null}
    id="texty"
    onChange={[Function]}
    readOnly={null}
    type="text"
  />
</label>
`;

exports[`TextInput has the correct width 2`] = `
<label
  className="control-label"
  htmlFor="texty"
  style={
    Object {
      "width": "50%",
    }
  }
>
  Texty
  <input
    className="form-control"
    disabled={null}
    id="texty"
    onChange={[Function]}
    readOnly={null}
    type="text"
  />
</label>
`;

exports[`TextInput renders 1`] = `
<label
  className="control-label"
  htmlFor="texty"
  style={
    Object {
      "width": null,
    }
  }
>
  Texty
  <input
    className="form-control"
    disabled={null}
    id="texty"
    onChange={[Function]}
    readOnly={null}
    type="text"
  />
</label>
`;

exports[`TextInput shows errors properly 1`] = `
<label
  className="control-label has-error"
  htmlFor="texty"
  style={
    Object {
      "width": null,
    }
  }
>
  Texty
  <input
    className="form-control"
    disabled={null}
    id="texty"
    onChange={[Function]}
    readOnly={null}
    type="text"
  />
</label>
`;

exports[`TextInput shows placeholder properly 1`] = `
<label
  className="control-label"
  htmlFor="texty"
  style={
    Object {
      "width": null,
    }
  }
>
  Texty
  <input
    className="form-control"
    disabled={null}
    id="texty"
    onChange={[Function]}
    placeholder="nice"
    readOnly={null}
    type="text"
  />
</label>
`;
